backend/tracim_backend/lib/webdav/dav_provider.py

```diff
@@ -1,33 +1,140 @@
 # coding: utf8
-from os.path import basename
-from os.path import dirname
+from collections import deque
+import enum
 import typing
 
 from pluggy import PluginManager
-from sqlalchemy.orm.exc import NoResultFound
 from wsgidav.dav_provider import DAVProvider
 from wsgidav.lock_manager import LockManager
 
-from tracim_backend.app_models.contents import content_type_list
 from tracim_backend.config import CFG
 from tracim_backend.exceptions import ContentNotFound
 from tracim_backend.exceptions import NotAuthenticated
 from tracim_backend.exceptions import WorkspaceNotFound
 from tracim_backend.lib.core.content import ContentApi
-from tracim_backend.lib.core.content import ContentRevisionRO
 from tracim_backend.lib.core.user import UserApi
 from tracim_backend.lib.core.workspace import WorkspaceApi
 from tracim_backend.lib.utils.request import TracimContext
 from tracim_backend.lib.utils.utils import normpath
 from tracim_backend.lib.utils.utils import webdav_convert_file_name_to_bdd
 from tracim_backend.lib.webdav import resources
 from tracim_backend.lib.webdav.lock_storage import LockStorage
+from tracim_backend.lib.webdav.resources import get_content_resource
+from tracim_backend.lib.webdav.resources import get_workspace_resource
+from tracim_backend.models.auth import User
 from tracim_backend.models.data import Content
-from tracim_backend.models.data import ContentNamespaces
 from tracim_backend.models.data import Workspace
 from tracim_backend.models.tracim_session import TracimSession
 
 
+class ResourceLevelType(enum.Enum):
+    ROOT = "root"
+    WORKSPACE = "workspace"
+    CONTENT = "content"
+
+
+class ProcessedWebdavPath(object):
+    """
+    Processor for Webdav Path:
+    - get all workspaces hierarchy from path
+    - get all content hierarchy from path
+    - provide useful properties to handle Webdav Request
+    """
+
+    def __init__(self, path: str, current_user: User, session: TracimSession, app_config: CFG):
+        self.path = path
+        self.workspace_api = WorkspaceApi(
+            current_user=current_user, session=session, config=app_config
+        )
+        self.content_api = ContentApi(current_user=current_user, session=session, config=app_config)
+
+        self.workspaces = deque()
+        self.contents = deque()
```

Comment:
Why do you use a `deque` here? Wouldn't a list be enough?

Reply:
We do need to : 
- add item at the end 
- get last item without removing it
nice to have : 
- be able to debug the code by just reading the list.

If we do use a simple list, it will work but as one of on first goal is to get the current last items, it's a bit more logical to have a list that allow direct access to the last item (without fetching all items). I do choose deque for this, but maybe there is better solution.

Reply:
I think that python's lists have a constant time random access operation (see https://wiki.python.org/moin/TimeComplexity) so they also would be OK I think.

Reply:
Interesting point. I do not have realize that python list are array.

Reply:
converted to list.